### PR TITLE
feat: migrate HTTP client from axios to Node.js fetch API

### PR DESCRIPTION
**Pull request by Kesin11**
## Summary

- Drop `axios` / `axios-retry` in response to recent axios vulnerability reports and move to the Node.js 24 native fetch API
- New `src/client/http_client.ts` provides `createHttpClient`, preserving the behaviour of the previous axios interceptors (debug request logging and error summarization) as built-in logic inside the factory
- Re-implements everything axios used to absorb implicitly: 3-attempt exponential-backoff retry, keepAlive/maxSockets via undici `Agent` + `setGlobalDispatcher`, `validateStatus`, `responseType: "arraybuffer"`, Basic auth, `params` `undefined`/`null` dropping, etc.
- Renames `summarizeAxiosError` → `summarizeHttpError` (same output shape, log consumers unaffected); `runner.ts` now uses `error instanceof HttpError`
- Migrates CircleCI v1/v2, Bitrise (keeps the two-instance setup for authed + S3), and Jenkins clients to `HttpClient`. GitHub is unchanged since it uses Octokit
- Adds `undici` to dependencies so we can configure the global dispatcher for `Agent`

## Test plan

- [x] `npm run check` passes (tsc)
- [x] `npx biome check .` passes (lint/format)
- [x] `npm test` — all 163 tests green, including the new `tests/client/http_client.test.ts` (15 cases)
- [x] `npm run build` — esbuild bundle succeeds
- [x] `node dist/index.mjs -c ci_analyzer.yaml --debug --only-services=github,circleci` — successfully fetches workflow/test data from CircleCI (~1.5 MB of output generated)

## axios vs fetch behaviour differences handled explicitly

- URL join: concatenate as strings (with slash-dedup) because `new URL(path, baseURL)` destroys the baseURL path segment
- Drop `undefined`/`null` query params (URLSearchParams keeps them otherwise)
- Regenerate `AbortSignal.timeout` per retry (aborted signals cannot be reused)
- Wrap fetch's network `TypeError` into `HttpError` so `runner.ts`'s type check still works
- Treat 204 / `Content-Length: 0` as `data: null` to avoid `response.json()` throwing on empty bodies
- Absolute URLs pass through and ignore baseURL (Bitrise S3 signed URLs, etc.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/client/artifact.ts
Comment (Kesin11) [inline src/client/artifact.ts:1]: This file will ultimately contain only type definitions, so I would like to move the type definitions to another file and delete this one.

File: src/client/http_client.ts [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+  params?: Record<string, unknown>;
+  headers?: Record<string, string>;
+  responseType?: "json" | "arraybuffer" | "text";
+  validateStatus?: (status: number) => boolean;
+  signal?: AbortSignal;
+  timeout?: number;
+};
+
+// biome-ignore lint/suspicious/noExplicitAny: axios compat — callers narrow via generic
+export type HttpResponse<T = any> = {
```
Comment (Kesin11) [inline src/client/http_client.ts]: Can replace `any` to `unknown` types?

File: src/client/http_client.ts [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+const isAbsoluteUrl = (url: string): boolean => /^https?:\/\//i.test(url);
+
+const joinUrl = (baseURL: string | undefined, url: string): string => {
+  if (isAbsoluteUrl(url)) return url;
+  if (!baseURL) return url;
+  return `${baseURL.replace(/\/+$/, "")}/${url.replace(/^\/+/, "")}`;
+};
+
+const buildQueryString = (params?: Record<string, unknown>): string => {
```
Comment (Kesin11) [inline src/client/http_client.ts]: This function is bit complicated. Please write JSDoc about this function behavior and also write unit tests.

File: tests/client/http_client.test.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        expect(req.url).toBe("/api/v1.1/project/github/owner/repo");
+        res.setHeader("content-type", "application/json");
+        res.end(JSON.stringify({ ok: true }));
+      };
+      const client = createHttpClient(logger, options, {
+        baseURL: `${origin}/api/v1.1`,
+      });
+      const res = await client.get("project/github/owner/repo");
+      expect(res.status).toBe(200);
+      expect(res.data).toEqual({ ok: true });
```
Comment (Kesin11) [inline tests/client/http_client.test.ts]: Use `expect.objectContaining` instead of multiple expect to same object.
ref: https://vitest.dev/api/expect.html#expect-objectcontaining

File: tests/client/http_client.test.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        await client.get("api", { params: { foo: "bar" } });
+        expect.fail("should throw");
+      } catch (e) {
+        expect(e).toBeInstanceOf(HttpError);
+        const err = e as HttpError;
+        expect(err.request.method).toBe("GET");
+        expect(err.request.url).toBe("api");
+        expect(err.request.baseURL).toBe(origin);
+        expect(err.request.params).toEqual({ foo: "bar" });
+        expect(err.response?.status).toBe(403);
```
Comment (Kesin11) [inline tests/client/http_client.test.ts]: Use `expect.objectContaining` instead of multiple expect to same object.
ref: https://vitest.dev/api/expect.html#expect-objectcontaining

File: src/client/artifact.ts
Comment (Kesin11) [reply src/client/artifact.ts:1]: Moved `Artifact` / `CustomReportArtifact` types to a new `src/client/artifact.ts` and deleted `client.ts` which had become a types-only stub. All import sites updated accordingly. (39e3067)

File: src/client/http_client.ts [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+        status: response.status,
+        statusText: response.statusText,
+        headers: response.headers,
+      };
+    } catch (error) {
+      lastError = error;
+      if (attempt <= MAX_RETRY && shouldRetry(error)) {
+        await sleep(exponentialBackoffDelay(attempt));
+        continue;
+      }
```
[the remark this answers is not in the document]
Comment (Kesin11) [reply src/client/http_client.ts:292]: Good catch! Added a `userAborted` parameter to `shouldRetry` and short-circuit retries immediately when `req.userSignal?.aborted` is `true`. Timeout-only aborts still retry as before. Also made the backoff `sleep` cancellable via the caller's `AbortSignal` so an abort during the backoff delay bails out without waiting for the sleep to finish. Added tests for both "user abort does not retry" and "timeout abort still retries". (39e3067)

File: src/client/bitrise_client.ts [hunk 6]
```diff
@@ -229,7 +225,7 @@ export class BitriseClient {
     const pathResponses = artifactResponses.map((artifact) => {
       // NOTE: Bitrise store their artifacts to S3.
       // S3 does not accept 'Authorization' header, so using another axios client that is not set Bitrise Authorization header.
```
[the remark this answers is not in the document]
Comment (Kesin11) [reply src/client/bitrise_client.ts]: Updated the comment to reference `#artifactHttp` (`HttpClient`) instead of "another axios client". (39e3067)

File: src/client/http_client.ts (hunk 1 above)
Comment (Kesin11) [reply src/client/http_client.ts]: Changed the default generic of `HttpResponse`, `HttpClient.get`, and `post` from `any` to `unknown`, and removed all `biome-ignore` annotations. Each call site now passes an explicit type via `this.#http.get<T>(...)`, making the redundant `as T` casts unnecessary. (39e3067)

File: src/client/http_client.ts (hunk 2 above)
Comment (Kesin11) [reply src/client/http_client.ts]: Exported `buildQueryString` and added JSDoc documenting its behaviour (dropping `undefined`/`null` values, array expansion into repeated keys, `String()` coercion, etc.). Also added 7 dedicated unit tests covering empty/undefined/null/array/encoding cases. (39e3067)

File: tests/client/http_client.test.ts (hunk 3 above)
Comment (Kesin11) [reply tests/client/http_client.test.ts]: Refactored to use `expect.objectContaining`. (39e3067)

File: tests/client/http_client.test.ts (hunk 4 above)
Comment (Kesin11) [reply tests/client/http_client.test.ts]: Refactored the `HttpError` assertion to use nested `expect.objectContaining` for both `request` and `response`. (39e3067)

**Comment by Kesin11** [comment]
Please remove `axios` from copilot-instructions.md
